lib/renderer/src/renderer/core/sokol/SokolTexture.cpp

```diff
@@ -0,0 +1,391 @@
+#include "../../exception.h"
+
+#define CODE(...) #__VA_ARGS__
+using namespace renderer::core;
+
+sg_shader get_shader_rgba();
+
+typedef struct {
+	hmm_mat4 transform;
+	hmm_vec4 uv_transform;
+} vs_params_t;
+
+typedef struct {
+	hmm_vec4 color;
+} fs_params_t;
+
+
+struct renderer::core::RenderContext {
+	sg_pipeline pip;
+	sg_bindings bind;
+	sg_buffer vertex_buffer;
+	sg_buffer index_buffer;
+	sg_shader shader;
+	sg_image image;
+};
+
+
+SokolTexture::SokolTexture(TextureType texture_type, BlendMode blend_mode, int32_t width, int32_t height,
+						   std::unique_ptr<RenderContext> &&render_context)
+	: _texture_type(texture_type)
+	, _blend_mode(blend_mode)
+	, _width(width)
+	, _height(height)
+	, _color({255, 255, 255, 255})
+	, _render_context(std::move(render_context))
+	, _data(nullptr)
+{}
+
+SokolTexture::~SokolTexture() {
+	delete[] _data;
+}
+
+std::unique_ptr<SokolTexture>
+SokolTexture::create(TextureType texture_type, BlendMode blend_mode, int32_t width, int32_t height) {
+	std::unique_ptr<RenderContext> render_context = std::make_unique<RenderContext>(RenderContext{});
+
+	float vertices[] = {
+		-1.0f,  1.0f, 0.0f, 0.0f,
+		 1.0f,  1.0f, 1.0f, 0.0f,
+		 1.0f, -1.0f, 1.0f, 1.0f,
+		-1.0f, -1.0f, 0.0f, 1.0f,
+	};
+	sg_buffer_desc vbuf_desc = {
+			.size = 0,
+			.type = SG_BUFFERTYPE_VERTEXBUFFER,
+			.data = SG_RANGE(vertices),
+	};
+	sg_buffer vbuf = sg_make_buffer(&vbuf_desc);
+	render_context->bind.vertex_buffers[0] = vbuf;
+	render_context->vertex_buffer = vbuf;
+
+	uint16_t indices[] = {
+			0, 1, 2,
+			0, 2, 3
+	};
+
+	sg_buffer_desc ibuf_desc = {
+//			.size = sizeof(indices),
+			.type = SG_BUFFERTYPE_INDEXBUFFER,
+			.data = SG_RANGE(indices),
+	};
+
+	sg_buffer ibuf = sg_make_buffer(ibuf_desc);
+	render_context->index_buffer = ibuf;
+	render_context->bind.index_buffer = ibuf;
+
+
+	sg_shader shd;
+
+	switch (texture_type) {
+		case TextureType::RGBA32:{
+			 sg_image image = sg_make_image({
+								.width = (int32_t)width,
+								.height = (int32_t)height,
+								.usage = SG_USAGE_DYNAMIC,
+								.pixel_format = SG_PIXELFORMAT_RGBA8,
+								.min_filter = SG_FILTER_LINEAR,
+								.mag_filter = SG_FILTER_LINEAR,
+							});
+			shd = get_shader_rgba();
+			render_context->bind.fs_images[0] = image;
+			render_context->image = image;
+		}
+		break;
+		default:
+			throw RendererException(std::string("InvalidTextureType ") + std::to_string((int)texture_type));
+	}
+
+	render_context->shader = shd;
+
+	sg_blend_state blend_state;
+	switch (blend_mode) {
+		case BlendMode::Alpha:
+			blend_state = {
+					.enabled = true,
+					.src_factor_rgb = SG_BLENDFACTOR_SRC_ALPHA,
+					.dst_factor_rgb = SG_BLENDFACTOR_ONE_MINUS_SRC_ALPHA,
+					.op_rgb = SG_BLENDOP_ADD,
+					.src_factor_alpha = SG_BLENDFACTOR_ONE,
+					.dst_factor_alpha = SG_BLENDFACTOR_ZERO,
+					.op_alpha = SG_BLENDOP_ADD,
+			};
+			break;
+		default:
+			blend_state = {};
+			break;
+	}
+	sg_pipeline pip = sg_make_pipeline({

```

Comment:
I find it peculiar that we are creating a pipeline as a part of the texture. Is this how people use Sokol generally?

Reply:
Uh, no that's just me copied the code from examples :)  I'll move it to the higher level